experimental/mrpc/plugin/policy/logging/src/engine.rs

```diff
@@ -0,0 +1,148 @@
+
+    pub(crate) indicator: Indicator,
+    pub(crate) config: LoggingConfig,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum Status {
+    Progress(usize),
+    Disconnected,
+}
+
+use Status::Progress;
+
+impl Engine for LoggingEngine {
+    fn activate<'a>(self: Pin<&'a mut Self>) -> BoxFuture<'a, EngineResult> {
+        Box::pin(async move { self.get_mut().mainloop().await })
+    }
+
+    fn description(self: Pin<&Self>) -> String {
+        "LoggingEngine".to_owned()
+    }
+
+    #[inline]
+    fn tracker(self: Pin<&mut Self>) -> &mut Indicator {
+        &mut self.get_mut().indicator
+    }
+
+    fn handle_request(&mut self, request: Vec<u8>, _cred: UCred) -> Result<()> {
+        let request: control_plane::Request = bincode::deserialize(&request[..])?;
+
+        match request {
+            control_plane::Request::NewConfig() => {
+                self.config = LoggingConfig {};
+            }
+        }
+        Ok(())
+    }
+}
+
+impl_vertex_for_engine!(LoggingEngine, node);
+
+impl Decompose for LoggingEngine {
+    fn flush(&mut self) -> Result<()> {
+        while !self.tx_inputs()[0].is_empty() {
+            self.check_input_queue()?;
+        }
+        Ok(())
+    }
+
+    fn decompose(
+        self: Box<Self>,
+        _shared: &mut SharedStorage,
+        _global: &mut ResourceCollection,
+    ) -> (ResourceCollection, DataPathNode) {
+        let engine = *self;
+
+        let mut collections = ResourceCollection::with_capacity(4);
+        collections.insert("config".to_string(), Box::new(engine.config));
+        (collections, engine.node)
+    }
+}
+
+impl LoggingEngine {
+    pub(crate) fn restore(
+        mut local: ResourceCollection,
+        node: DataPathNode,
+        _prev_version: Version,
+    ) -> Result<Self> {
+        let config = *local
+            .remove("config")
+            .unwrap()
+            .downcast::<LoggingConfig>()
+            .map_err(|x| anyhow!("fail to downcast, type_name={:?}", x.type_name()))?;
+
+        let engine = LoggingEngine {
+            node,
+            indicator: Default::default(),
+            config,
+        };
+        Ok(engine)
+    }
+}
+
+impl LoggingEngine {
+    async fn mainloop(&mut self) -> EngineResult {
+        loop {
+            let mut work = 0;
+            // check input queue, ~100ns
+            loop {
+                match self.check_input_queue()? {
+                    Progress(0) => break,
+                    Progress(n) => work += n,
+                    Status::Disconnected => return Ok(()),
+                }
+            }
+            // If there's pending receives, there will always be future work to do.
+            self.indicator.set_nwork(work);
+
+            future::yield_now().await;
+        }
+    }
+}
+
+impl LoggingEngine {
+    fn check_input_queue(&mut self) -> Result<Status, DatapathError> {
+        use phoenix_common::engine::datapath::TryRecvError;
+
+        match self.tx_inputs()[0].try_recv() {
```

Comment:
I'm wondering if you need to also `try_recv` from `rx_inputs()[0]` and forward everything received to `rx_outputs()`. Because I see in [eval/policy/logging/attach.toml](https://github.com/phoenix-dataplane/phoenix/pull/233/files#diff-f245724397adaae2f6cb908809d9526f3e195ba401b13b536d1ef06a357a08b4) you replaced both rx and tx channels.

Reply:
ah, you're right. For testing, I only used the tx channels. 

This PR is meant to write a skeleton code for the logging engine. We will add more logic to it later. 

Perhaps we can merge this first?

Reply:
okay